nixos-generate-config: Add nixpkgs.hostPlatform to hardware-configuration.nix

roberth opened the pull request:
###### Description of changes

Resolves #177664 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos-generate-config: Add nixpkgs.system to hardware-config.nix

lovesegfault commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -84,6 +84,15 @@ sub debug {
+    die "Failed to retrieve current system type from nix.\n";
+}
+my $system = substr(@systemLines[0], 0, -1);
+push @attrs, "nixpkgs.system = lib.mkDefault $system;";
```
I think it's more correct to set `localSystem`?

roberth replied:
Hm, you have a point, but `localSystem` isn't correct either, when `crossSystem` is set.
So the thing is, we can't really specify the host system with a single option.

What we want to set is the "host" system.

|Variable |Meaning when straight  | Meaning when cross|
--- | --- | ---|
|system|host|build|
|localSystem|host|build|
|crossSystem|null|host|

So there's no variable for `host`.

Damn, for a minute I thought this change was going to be simple.

lovesegfault replied:
I'd be fine with `nixos-generate-config` assuming you're _not_ cross-compiling. It can also add a comment to the gen'd code saying "If you're cross-compiling, make sure to set crossSystem", or something like that.

roberth replied:
Ok, I've added new options.

|Variable |Meaning when straight  | Meaning when cross|
--- | --- | ---|
|hostSystem|host|host|
|buildSystem|build (=host)|build|
|LEGACY system|host|build|
|LEGACY localSystem|host|build|
|LEGACY crossSystem|null|host|

roberth replied:
Didn't get your message before my previous comment somehow.

> I'd be fine with `nixos-generate-config` assuming you're _not_ cross-compiling. It can also add a comment to the gen'd code saying "If you're cross-compiling, make sure to set crossSystem", or something like that.

The existing options are too complicated and less meaningful than the general / Nixpkgs terminology. We ought to fix that.

I appreciate that you try to reduce scope creep, but in this case doing half the work results in the generation of half-assed config state that resides on user machines. `hardware-configuration.nix` should either be a legacy one without any `nixpkgs.*` options, or one with `nixpkgs.hostPlatform`. Supporting a third variation would be painful.

Ericson2314 replied:
Do you mean to change this line after everything else?

lovesegfault replied:
> I appreciate that you try to reduce scope creep, but in this case doing half the work results in the generation of half-assed config state that resides on user machines. hardware-configuration.nix should either be a legacy one without any nixpkgs.* options, or one with nixpkgs.hostPlatform. Supporting a third variation would be painful.

I was mostly trying to unblock you, without asking you to do a bunch of extra work :)

> The existing options are too complicated and less meaningful than the general / Nixpkgs terminology. We ought to fix that.

I agree, excited about the changes you made!

roberth committed:
nixos: Add simplified nixpkgs.{hostSystem,buildSystem}

Ericson2314 commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -157,6 +175,46 @@ in
+    buildSystem = mkOption {
+      type = types.either types.str types.attrs; # TODO utilize lib.systems.parsedPlatform
+      default = cfg.hostSystem;
+      example = "x86_64-linux";
```
```suggestion
      example = { system = "aarch64-linux"; config = "aarch64-unknown-linux-gnu"; };
```

roberth replied:
Done, but I picked `x86_64` because it's more "dev machine"-like, at least for now.

Ericson2314 commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -55,9 +55,27 @@ let
+    then
+      let isCross = cfg.buildPlatform != cfg.hostPlatform;
+          systemArgs =
+            if isCross == builtins.trace opt.buildPlatform true
```
Still want the trace?

Ericson2314 commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -244,6 +304,21 @@ in
+      {
+        # soft deprecate crossSystem when hostPlatform is set
+        assertion = opt.hostPlatform.isDefined -> !opt.crossSystem.isDefined;
+        message = "If you set hostPlatform, you must only define hostPlatform and/or buildPlatform, but none of the legacy options such as crossSystem.";
```
```
let explanation = "If you set hostPlatform, you must only define hostPlatform and/or buildPlatform, but none of the legacy options (localSystem crossSystem or system)";
```
```suggestion
        message = ''do not set crossSystem with hostPlatform. ${explanation}";
```

Ericson2314 replied:
And shouldn't we do for buildPlatform too?

roberth replied:
Done. It's a single assertion message now. See `test.nix`.

roberth committed:
lib/options: Add showOptionWithDefLocs

roberth committed:
nixos-generate-config: nixpkgs.system -> nixpkgs.hostPlatform

roberth commented:
@ofborg test installer

lovesegfault reviewed (approved):
This looks good, a couple of questions:

1. Shouldn't `localSystem` get deprecated too?
2. Add a `lib.warn` to the deprecated items?
3. Add something to the release notes?

roberth committed:
release-notes: Add nixpkgs.hostPlatform changes

roberth commented:
> 1. Shouldn't `localSystem` get deprecated too?

Kind of. It's bunched in with the other "legacy" options in the assertion the enforces the mutual exclusion of the old and new options.

> 2. Add a `lib.warn` to the deprecated items?

Not yet. The new options need to exist for some time to make it feasible for the ecosystem to start a migration. For now it's only a preview for flakes users and "typical" NixOS users. Deployment tools like NixOps and the numerous other ones can only switch when the new options are supported in all releases. Adding warnings will only upset people.

> 3. Add something to the release notes?

Ah yes, of course. Done.

-----

The manual does not need changes yet, because the old way is fully supported. Changes can be made if and when it is decided later that this becomes the blessed way.

lovesegfault commented:
Sounds good, thanks for this!

roberth committed:
nixos-generate-config: Make robust against missing newline

The substr solution assumed a newline to be present.
The new solution will not remove the newline if it goes missing in the future.

Apparently this is idiomatic perl.

Thanks pennae for the suggestion!

roberth commented:
@ofborg test installer

K900 commented:
Tests are broken now: https://hydra.nixos.org/build/181556482/nixlog/1

roberth commented:
What? I ran those tests on ofborg! I guess I'll fix them.

roberth commented:
This should help with the channel blocker, although ofborg is broken and I can't run the tests locally.
https://github.com/NixOS/nixpkgs/pull/178958

roberth commented:
Fix confirmed: https://hydra.nixos.org/build/181594825
